components/post_view/reaction/reaction.jsx

```diff
@@ -83,15 +85,155 @@ export default class Reaction extends React.PureComponent {
         sortedUsers: PropTypes.object.isRequired,
     }
 
-    handleAddReaction = (e) => {
-        e.preventDefault();
-        const {actions, post, emojiName} = this.props;
-        actions.addReaction(post.id, emojiName);
+    constructor(props) {
+        super(props);
+
+        const {reactionCount, canRemoveReaction, canAddReaction} = this.props;
+        const {currentUserReacted} = this.props.sortedUsers;
+
+        this.state = {
+            canAddReaction,
+            canRemoveReaction,
+        };
+
+        if (currentUserReacted) {
+            this.state = {
+                ...this.state,
+                userReacted: currentUserReacted,
+                reactedClass: 'Reaction--reacted',
+                displayNumber: reactionCount,
+                reactedNumber: reactionCount,
+                unreactedNumber: reactionCount - 1,
+            };
+        } else {
+            this.state = {
+                ...this.state,
+                userReacted: currentUserReacted,
+                reactedClass: 'Reaction--unreacted',
+                displayNumber: reactionCount,
+                reactedNumber: reactionCount + 1,
+                unreactedNumber: reactionCount,
+            };
+        }
+
+        if (!canAddReaction || !canRemoveReaction) {
+            this.state = {
+                ...this.state,
+                reactedClass: 'Reaction--read-only',
+            };
+        }
+
+        this.reactionButtonRef = React.createRef();
+        this.reactionCountRef = React.createRef();
+        this.reactedNumeralRef = React.createRef();
+    }
+
+    static getDerivedStateFromProps(nextProps, prevState) {
+        // reaction count has changed, but not by current user
+        if (nextProps.reactionCount !== prevState.displayNumber && nextProps.sortedUsers.currentUserReacted === prevState.userReacted) {
+            // set counts relative to current user having reacted
+            if (prevState.userReacted) {
+                return {
+                    displayNumber: nextProps.reactionCount,
+                    reactedNumber: nextProps.reactionCount,
+                    unreactedNumber: nextProps.reactionCount - 1,
+                };
+            }
+
+            // set counts relative to current user having NOT reacted
+            return {
+                displayNumber: nextProps.reactionCount,
+                reactedNumber: nextProps.reactionCount + 1,
+                unreactedNumber: nextProps.reactionCount,
+            };
+        }
+
+        // if permission to add reactions has changed
+        if (nextProps.canAddReaction !== prevState.canAddReaction) {
+            // if can't add reaction
+            if (!nextProps.canAddReaction) {
+                return {
+                    reactedClass: 'Reaction--read-only',
```

Comment:
Why store the read only class in the state? Couldn't we apply both the `Reaction--reacted` and `Reaction--read-only` classes to the component at the same time and have the `read-only` one take precedence? If we do that, we'll have less stuff to track in the component state.

Reply:
I understand what you're saying, because I was also little confused around the design intent (trying to understand from the CSS). It seems there is only one "read-only" style that doesn't discriminate between whether the current user has reacted or not. I can make this change and see if it still passes the use-cases for QA.

Reply:
Updated per our discussion @hmhealey